presto-hive/src/main/java/com/facebook/presto/hive/CachingDirectoryLister.java

```diff
@@ -90,19 +91,20 @@ public Iterator<HiveFileInfo> list(
             List<HiveFileInfo> files = cache.getIfPresent(path);
             if (files != null) {
                 runtimeStats.addMetricValue(DIRECTORY_LISTING_CACHE_HIT, NONE, 1);
+                runtimeStats.addMetricValue(FILES_READ_COUNT, NONE, files.size());
                 return files.iterator();
             }
         }
 
         runtimeStats.addMetricValue(DIRECTORY_LISTING_CACHE_MISS, NONE, 1);
         Iterator<HiveFileInfo> iterator = delegate.list(fileSystem, table, path, partition, namenodeStats, hiveDirectoryContext);
         if (hiveDirectoryContext.isCacheable() && cachedTableChecker.isCachedTable(table.getSchemaTableName())) {
-            return cachingIterator(iterator, path);
+            return metricIteratorWithCache(iterator, Optional.ofNullable(path), runtimeStats, true);
         }
-        return iterator;
+        return metricIteratorWithCache(iterator, Optional.empty(), runtimeStats, false);
     }
 
-    private Iterator<HiveFileInfo> cachingIterator(Iterator<HiveFileInfo> iterator, Path path)
+    private Iterator<HiveFileInfo> metricIteratorWithCache(Iterator<HiveFileInfo> iterator, Optional<Path> path, RuntimeStats runtimeStats, boolean isCached)
```

Comment:
why is path made optional ?
also change `isCached` --> `enableCaching`

Reply:
We only need path when `enableCaching` is set to true because it is the key for cache, but I can change it so we pass path in both cases if that is better.